scout/parse/models.py

```diff
@@ -0,0 +1,419 @@
+"""Class to hold information about scout load config and subclasses used"""
+
+import datetime
+import logging
+import re
+from fractions import Fraction
+from glob import glob
+from pathlib import Path
+from typing import Any, ByteString, Dict, List, Optional, Union
+
+from pydantic import BaseModel, Field, root_validator, validator
+from typing_extensions import Literal
+
+from scout.exceptions import ConfigError, PedigreeError
+from scout.utils.date import get_date
+
+LOG = logging.getLogger(__name__)
+
+# As the class constructor is called twice, this messes with the
+# aliases, removing the set value the second time called. It seems
+# to work by adding the aliased name as a attribute...
+#
+class ChromographImages(BaseModel):
+    autozygous: Optional[str] = None
+    coverage: Optional[str] = None
+    upd_regions: Optional[str] = None
+    upd_sites: Optional[str] = None
+
+
+def _glob_wildcard(path):
+    """Search for multiple files using a path with wildcard."""
+    wildcard = re.search(r"{([A-Za-z0-9_-]+)}", path)
+    # make proper wildcard path
+    glob_path = path[: wildcard.start()] + "*" + path[wildcard.end() :]
+    wildcard_end = len(path) - wildcard.end()
+    paths = tuple(
+        {
+            "match": match[wildcard.start() : -wildcard_end],
+            "variable_name": wildcard.group(1),
+            "path": Path(match),
+        }
+        for match in glob(glob_path)
+    )
+    return paths
+
+
+def _replace_wildcard_with_match(match, image):
+    path_expanded = image.path.replace("{%s}" % match["variable_name"], match["match"])
+    str_repid = match["match"]
+    title_expanded = image.title.replace("{%s}" % match["variable_name"], match["match"])
+
+    return Image(
+        path=path_expanded,
+        description=image.description,
+        height=image.height,
+        format=image.format,
+        width=image.width,
+        title=title_expanded,
+        str_repid=str_repid,
+    )
+
+
+def is_wildcard(string_path):
+    """Return true if string_path contains { and }, used to contruct wildcard matching"""
+    return "{" in string_path and "}" in string_path
+
+
+class Image(BaseModel):
+    data: Optional[str] = None
+    description: Optional[str] = None
+    height: Optional[int] = None
+    format: Optional[str] = None
+    path: str = None
+    str_repid: Optional[str] = None
+    title: str = None
+    width: Optional[int] = None
+
+    def __eq__(self, other):
+        return self.title == other.title
+
+    @root_validator
+    def valid_image_suffix(cls, values):
+        path = values.get("path")
+        path = Path(path)
+
+        # Skip configured image if path suffix is not an image file type
+        if not path.suffix[1:] in ["gif", "svg", "png", "jpg", "jpeg"]:
+            values["path"] = None
```

Comment:
Looks like the idea is skipping the load of a custom image if the format is not standard (["gif", "svg", "png", "jpg", "jpeg"]). But what happens if you try to load a wrong image format instead ,is that the load will crash with this error:

```
2021-11-15 10:36:45 ChiaraRMBP.local scout.parse.models[55318] DEBUG IMAGE:data=None description='A very good description' height=100 format=None path=None str_repid='AFF2' title='A png image' width=500
Traceback (most recent call last):
  File "/Users/chiararasi/miniconda3/envs/py38/bin/scout", line 33, in <module>
    sys.exit(load_entry_point('scout-browser', 'console_scripts', 'scout')())
  File "/Users/chiararasi/miniconda3/envs/py38/lib/python3.8/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/Users/chiararasi/miniconda3/envs/py38/lib/python3.8/site-packages/flask/cli.py", line 600, in main
    return super().main(*args, **kwargs)
  File "/Users/chiararasi/miniconda3/envs/py38/lib/python3.8/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/Users/chiararasi/miniconda3/envs/py38/lib/python3.8/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/chiararasi/miniconda3/envs/py38/lib/python3.8/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/chiararasi/miniconda3/envs/py38/lib/python3.8/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/chiararasi/miniconda3/envs/py38/lib/python3.8/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/Users/chiararasi/miniconda3/envs/py38/lib/python3.8/site-packages/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/commands/setup/setup_scout.py", line 192, in demo
    setup_scout(
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/load/setup.py", line 202, in setup_scout
    config_data = parse_case_data(config=case_data)
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/parse/case.py", line 52, in parse_case_data
    config_dict = parse_case_config(config)
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/parse/case.py", line 114, in parse_case_config
    parsed_config = ScoutLoadConfig(**config)
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/parse/models.py", line 312, in __init__
    super().__init__(vcf_files=vcfs, **data)
  File "pydantic/main.py", line 406, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 3 validation errors for ScoutLoadConfig
custom_images -> __root__
  argument of type 'NoneType' is not iterable (type=type_error)
custom_images -> __root__
  expected str, bytes or os.PathLike object, not NoneType (type=type_error)
custom_images -> __root__
  expected str, bytes or os.PathLike object, not NoneType (type=type_error)
```

Which I think is also why the test `tests/parse/test_parse_case.py::test_parse_incorrect_custom_images` is failing at the moment.